services/federator/test/unit/Test/Federator/Validation.hs

```diff
@@ -76,224 +76,167 @@ tests =
-          runValidation $ validateDomain (Just localhostExampleCert) (domainText domain)
-    resFirst @?= Right domain
-    let resSecond :: Either InwardError Domain =
-          runValidation $ validateDomain (Just secondExampleCert) (domainText domain)
-    resSecond @?= Right domain
-
+    resFirst <-
+      runValidation $
+        validateDomain (Just localhostExampleCert) (toByteString' domain)
+    resFirst @?= domain
+    resSecond <-
+      runValidation $
+        validateDomain (Just secondExampleCert) (toByteString' domain)
+    resSecond @?= domain
+
+-- FUTUREWORK: is this test really necessary?
 validateDomainDiscoveryFailed :: TestTree
 validateDomainDiscoveryFailed =
   testCase "should fail if discovery fails" $ do
     exampleCert <- BS.readFile "test/resources/unit/example.com.pem"
-    let res :: Either InwardError Domain =
-          run . Polysemy.runError
-            . mockDiscoveryFailure
-            . Polysemy.runReader noClientCertSettings
-            $ validateDomain (Just exampleCert) "example.com"
-    res @?= Left (InwardError IDiscoveryFailed "DNS error: mock DNS error")
+    res <-
+      runM . runError
+        . assertNoError @ValidationError
+        . mockDiscoveryFailure
+        . runInputConst noClientCertSettings
+        $ validateDomain (Just exampleCert) "example.com"
+    res @?= Left (DiscoveryFailureDNSError "mock DNS error")
 
 validateDomainNonIdentitySRV :: TestTree
 validateDomainNonIdentitySRV =
   testCase "should run discovery to look up the federator domain" $ do
     exampleCert <- BS.readFile "test/resources/unit/localhost.example.com.pem"
     let domain = Domain "foo.example.com"
-        res :: Either InwardError Domain =
-          run . Polysemy.runError
-            . mockDiscoveryMapping domain ("localhost.example.com" :| [])
-            . Polysemy.runReader noClientCertSettings
-            $ validateDomain (Just exampleCert) (domainText domain)
-    res @?= Right domain
-
-validatePathSuccess :: [TestTree]
-validatePathSuccess = do
-  let paths =
-        [ "federation/get-user-by-handle",
-          "federation/get-conversations"
-        ]
-  expectOk <$> paths
-  where
-    expectOk :: ByteString -> TestTree
-    expectOk path = testCase ("should allow " <> cs path) $ do
-      runSanitize path @?= Right path
-
-validatePathNormalize :: [TestTree]
-validatePathNormalize = do
-  let paths =
-        [ ("federation//stuff", "federation/stuff"),
-          ("/federation/get-user-by-handle", "federation/get-user-by-handle"),
-          ("federation/../federation/stuff", "federation/stuff")
-        ]
-  expectNormalized <$> paths
-  where
-    expectNormalized :: (ByteString, ByteString) -> TestTree
-    expectNormalized (input, output) = do
-      testCase ("Should allow " <> cs input <> " and normalize to " <> cs output) $ do
-        runSanitize input @?= Right output
-
-validatePathForbidden :: [TestTree]
-validatePathForbidden = do
```

Comment:
Could you leave these tests (or tests similar in spirit) in? Assuming some other component already validated a path so that disallowed paths cannot be called is not a good assumption I'd say. What prevents me from calling federator with /federation/brig/<double percent encoded form of /i/delete-user> where maybe one set of encoding is removed from e.g. the ingress, but another encoding makes its way through?

Maybe there is no way this can happen, but then these tests should do no harm.

The alternative would be to validate incoming paths to a known list of path segments for that component, e.g. extract all the /federation/some-path segments out from the servant API and ensure incoming requests are an exact match of one of those.

At the moment, the code blindly trusts this text component after /federation/ to be valid, as long as it's not empty. That leaves a risk that someone can craft some weird path allowing an external to use an otherwise-unreachable path brig or galley exposes. 

Am I holding things wrong?

Reply:
I think it's highly unlikely that there could be any attack like this, now that the RPC is just a path segment of the initial request, but we did leave most of those tests around to be extra safe. They are in `testInvalidPaths` in the `ExternalService` tests.

Here's what we removed (roughly):

 - Fragments being disallowed: this is slightly annoying because the Wai stuff does not handle fragments appropriately (it just treats `#` as a normal character). Anyway, they are innocuous AFAICT.
 - Length limits This will happen at some earlier point anyway.
 - ASCII only. There is no harm in allowing non-ASCII characters. They are roundtripped as UTF-8, and they will just result in a 404 downstream.

Path traversals via percent encoding are still tested, but the corresponding paths are allowed instead of being rejected. They get percent encoded again before being forwarded, so they are safe.

I'm actually wondering if there is actually any logic anywhere that interprets `..` as parent directory, enabling traversals. I don't think Wai or Servant do this. Were any path traversal attacks ever possible at any point?

Reply:
I added a validation step that enforces characters on rpc name in https://github.com/wireapp/wire-server/pull/1930/commits/356c8480b14819977a56a7d921dbcec9e85007fe